File: tests/test_kinetics_base.py
"""
Python integration tests for the KINETICS-base reader.

Tests the Python bindings for:
- KineticsOptions.from_kinetics_base()
- Kinetics module with KINETICS-base data
"""

import os
import pytest
import torch

torch.set_default_dtype(torch.float64)

DATA_DIR = os.path.join(os.path.dirname(__file__), "kinetics_base", "data")


@pytest.fixture
def master_path():
    return os.path.join(DATA_DIR, "test_master.inp")


@pytest.fixture
def catalog_path():
    return os.path.join(DATA_DIR, "test_catalog.dat")


@pytest.fixture
def cross_dir():
    return os.path.join(DATA_DIR, "cross") + "/"


def _count_reversible(reactions):
    """Count reversible reactions by checking equation string for <=>."""
    return sum(1 for r in reactions if "<=>" in r.equation())


def test_from_kinetics_base_no_xsec(master_path):
    """Load KINETICS-base master input without cross-sections."""
    import kintera as kt

    opts = kt.KineticsOptions.from_kinetics_base(master_path)

    species = opts.species()
    assert len(species) == 10
    assert "O" in species
    assert "O2" in species
    assert "O3" in species
    assert "H2O" in species
    assert "N2" in species
    assert "O(1D)" in species

    reactions = opts.reactions()
    assert len(reactions) == 20  # 12 thermal + 8 photolysis

    n_rev = _count_reversible(reactions)
    assert n_rev == 12  # all thermal reactions are reversible


def test_from_kinetics_base_with_xsec(master_path, catalog_path, cross_dir):
    """Load KINETICS-base master input with cross-sections."""
    import kintera as kt

    opts = kt.KineticsOptions.from_kinetics_base(
        master_path, catalog_path, cross_dir
    )

    reactions = opts.reactions()
    assert len(reactions) == 20


def test_kinetics_module_creation(master_path, catalog_path, cross_dir):
    """Create Kinetics module from KINETICS-base data."""
    import kintera as kt

    opts = kt.KineticsOptions.from_kinetics_base(
        master_path, catalog_path, cross_dir
    )
    kinet = kt.Kinetics(opts)
    assert kinet is not None

    species = opts.species()
    nspecies = len(species)
    nrxn = len(opts.reactions())
    n_rev = _count_reversible(opts.reactions())

    stoich = kinet.stoich
    assert stoich.size(0) == nspecies
    assert stoich.size(1) == nrxn + n_rev


def test_kinetics_forward_with_xsec(master_path, catalog_path, cross_dir):
    """Forward pass with cross-section data loaded."""
    import kintera as kt

    opts = kt.KineticsOptions.from_kinetics_base(
        master_path, catalog_path, cross_dir
    )
    kinet = kt.Kinetics(opts)

    species = opts.species()
    nspecies = len(species)

    temp = 300.0 * torch.ones(1)
    pres = 1.0e5 * torch.ones(1)
    conc = 1e18 * torch.ones(1, nspecies)

    wave = torch.linspace(100, 300, 10)
    aflux = 1e14 * torch.ones(10)
    extra = {"wavelength": wave, "actinic_flux": aflux}

    rate, rc_ddC, rc_ddT = kinet.forward(temp, pres, conc, extra)

    nrxn = len(opts.reactions())
    n_rev = _count_reversible(opts.reactions())
    assert rate.dim() == 2
    assert rate.size(0) == 1
    assert rate.size(1) == nrxn + n_rev

    du = rate @ kinet.stoich.t()
    assert du.size(-1) == nspecies


def test_kinetics_species_consistency(master_path):
    """Check that species names and weights are consistent."""
    import kintera as kt

    opts = kt.KineticsOptions.from_kinetics_base(master_path)

    species = opts.species()
    weights = kt.species_weights()
    assert len(species) == len(weights)
    assert all(w > 0 for w in weights)

    idx_o = species.index("O")
    assert abs(weights[idx_o] - 16.0) < 0.1

    idx_o2 = species.index("O2")
    assert abs(weights[idx_o2] - 32.0) < 0.1


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/test_kinetics_base.cpp
// external
#include <gtest/gtest.h>

// torch
#include <torch/torch.h>

// kintera
#include <configure.h>
#include <kintera/kinetics/kinetics.hpp>
#include <kintera/kinetics/kinetics_base_reader.hpp>
#include <kintera/kinetics/kinetics_formatter.hpp>

// tests
#include "device_testing.hpp"

using namespace kintera;

namespace kintera {
extern bool species_initialized;
}

static std::string data_dir() {
  return std::string(KINTERA_ROOT_DIR) +
         "/tests/kinetics_base/data/";
}

TEST(KineticsBaseParser, ParseMasterInput) {
  auto master = parse_kinetics_base_master(data_dir() + "test_master.inp");

  // Check elements
  EXPECT_EQ(master.elements.size(), 4);
  EXPECT_NEAR(master.elements.at("O"), 16.00, 0.01);
  EXPECT_NEAR(master.elements.at("H"), 1.01, 0.01);
  EXPECT_NEAR(master.elements.at("N"), 14.01, 0.01);
  EXPECT_NEAR(master.elements.at("C"), 12.01, 0.01);

  // Check species
  EXPECT_GE(master.species.size(), 10);

  // Find O2
  auto it = std::find_if(master.species.begin(), master.species.end(),
                          [](auto const& s) { return s.name == "O2"; });
  ASSERT_NE(it, master.species.end());
  EXPECT_EQ(it->composition.at("O"), 2);
  EXPECT_NEAR(it->molecular_weight, 32.0, 0.1);
  EXPECT_GE(it->n_nasa9_ranges, 2);

  // Find O(1D)
  auto it2 = std::find_if(master.species.begin(), master.species.end(),
                           [](auto const& s) { return s.name == "O(1D)"; });
  ASSERT_NE(it2, master.species.end());
  EXPECT_EQ(it2->composition.at("O"), 1);
  EXPECT_NEAR(it2->hf_kcal, 104.9, 0.1);

  // Check N2
  auto it3 = std::find_if(master.species.begin(), master.species.end(),
                           [](auto const& s) { return s.name == "N2"; });
  ASSERT_NE(it3, master.species.end());
  EXPECT_EQ(it3->composition.at("N"), 2);

  std::cout << "Species: " << master.species.size() << std::endl;
  for (auto const& sp : master.species) {
    std::cout << "  " << sp.name << " (MW=" << sp.molecular_weight
              << ", HF=" << sp.hf_kcal << ", NT=" << sp.n_nasa9_ranges
              << ")" << std::endl;
  }

  // Check reactions
  std::cout << "Photolysis: " << master.photolysis.size() << std::endl;
  for (auto const& rxn : master.photolysis) {
    std::cout << "  ";
    for (auto const& r : rxn.reactants) std::cout << r << " + ";
    std::cout << "=> ";
    for (auto const& p : rxn.products) std::cout << p << " + ";
    std::cout << std::endl;
  }

  std::cout << "Thermal: " << master.thermal.size() << std::endl;
  for (auto const& rxn : master.thermal) {
    std::string type =
        rxn.has_M ? (rxn.has_kinf ? "falloff" : "three-body") : "arrhenius";
    std::cout << "  [" << type << "] ";
    for (auto const& r : rxn.reactants) std::cout << r << " + ";
    std::cout << "=> ";
    for (auto const& p : rxn.products) std::cout << p << " + ";
    if (rxn.has_M) {
      std::cout << " (A0=" << rxn.A0 << ", b0=" << rxn.b0
                << ", Ea_R0=" << rxn.Ea_R0 << ")";
      if (rxn.has_kinf)
        std::cout << " (Ainf=" << rxn.A_inf << ", binf=" << rxn.b_inf << ")";
    } else {
      std::cout << " (A=" << rxn.A << ", b=" << rxn.b
                << ", Ea_R=" << rxn.Ea_R << ")";
    }
    std::cout << std::endl;
  }

  EXPECT_GE(master.photolysis.size(), 5);
  EXPECT_GE(master.thermal.size(), 5);

  // Count reaction types
  int n_arrhenius = 0, n_three_body = 0, n_falloff = 0;
  for (auto const& rxn : master.thermal) {
    if (rxn.has_M) {
      if (rxn.has_kinf)
        ++n_falloff;
      else
        ++n_three_body;
    } else {
      ++n_arrhenius;
    }
  }
  std::cout << "  Arrhenius: " << n_arrhenius << std::endl;
  std::cout << "  Three-body: " << n_three_body << std::endl;
  std::cout << "  Falloff: " << n_falloff << std::endl;

  EXPECT_GE(n_arrhenius, 1);
  EXPECT_GE(n_three_body, 1);
  EXPECT_GE(n_falloff, 1);
}

TEST(KineticsBaseParser, ParseCatalog) {
  auto catalog = parse_kinetics_base_catalog(data_dir() + "test_catalog.dat");
  EXPECT_GE(catalog.size(), 3);

  std::cout << "Catalog entries:" << std::endl;
  for (auto const& [eq, fname] : catalog) {
    std::cout << "  " << eq << " -> " << fname << std::endl;
  }
}

TEST(KineticsBaseParser, ParseCrossSection) {
  auto csf = parse_kinetics_base_cross_section(
      data_dir() + "cross/CROSS_O2=O2_LORES1.DAT");

  EXPECT_FALSE(csf.datasets.empty());
  if (!csf.datasets.empty()) {
    EXPECT_EQ(csf.datasets[0].type, 0);
    EXPECT_GT(csf.datasets[0].wavelengths_nm.size(), 10);

    std::cout << "Cross-section: " << csf.equation << std::endl;
    std::cout << "  Datasets: " << csf.datasets.size() << std::endl;
    std::cout << "  Type: " << csf.datasets[0].type << std::endl;
    std::cout << "  N wavelengths: " << csf.datasets[0].wavelengths_nm.size()
              << std::endl;
    if (!csf.datasets[0].wavelengths_nm.empty()) {
      std::cout << "  First wavelength (nm): "
                << csf.datasets[0].wavelengths_nm.front() << std::endl;
      std::cout << "  Last wavelength (nm): "
                << csf.datasets[0].wavelengths_nm.back() << std::endl;
    }
  }

  // Check branching ratio file
  auto csf2 = parse_kinetics_base_cross_section(
      data_dir() + "cross/CROSS_O2=2O_LORES1.DAT");
  EXPECT_FALSE(csf2.datasets.empty());
  if (!csf2.datasets.empty()) {
    EXPECT_EQ(csf2.datasets[0].type, 2);
    std::cout << "Cross-section (branching): " << csf2.equation << std::endl;
    std::cout << "  Type: " << csf2.datasets[0].type << std::endl;
    std::cout << "  N wavelengths: " << csf2.datasets[0].wavelengths_nm.size()
              << std::endl;
  }
}

TEST_P(DeviceTest, KineticsBaseLoadNoXsec) {
  // Reset species state
  species_initialized = false;

  auto op = KineticsOptionsImpl::from_kinetics_base(
      data_dir() + "test_master.inp", "", "", true);

  ASSERT_NE(op, nullptr);

  auto reactions = op->reactions();
  std::cout << "Total reactions: " << reactions.size() << std::endl;
  for (auto const& rxn : reactions) {
    std::cout << "  " << rxn.equation() << std::endl;
  }

  EXPECT_GT(op->arrhenius()->reactions().size(), 0);
  EXPECT_GT(op->three_body()->reactions().size(), 0);
  EXPECT_GT(op->lindemann_falloff()->reactions().size(), 0);
  EXPECT_GT(op->photolysis()->reactions().size(), 0);

  Kinetics kinet(op);
  kinet->to(device, dtype);

  std::cout << "Kinetics module created successfully" << std::endl;
  std::cout << "Stoich shape: " << kinet->stoich.sizes() << std::endl;
}

TEST_P(DeviceTest, KineticsBaseLoadWithXsec) {
  species_initialized = false;

  auto op = KineticsOptionsImpl::from_kinetics_base(
      data_dir() + "test_master.inp", data_dir() + "test_catalog.dat",
      data_dir() + "cross/", true);

  ASSERT_NE(op, nullptr);

  std::cout << "Photolysis reactions: " << op->photolysis()->reactions().size()
            << std::endl;
  std::cout << "Wavelength grid size: " << op->photolysis()->wavelength().size()
            << std::endl;
  std::cout << "Cross-section data size: "
            << op->photolysis()->cross_section().size() << std::endl;
}

TEST_P(DeviceTest, KineticsBaseForward) {
  species_initialized = false;

  auto op = KineticsOptionsImpl::from_kinetics_base(
      data_dir() + "test_master.inp", data_dir() + "test_catalog.dat",
      data_dir() + "cross/");

  Kinetics kinet(op);
  kinet->to(device, dtype);

  auto species = op->species();
  int nspecies = species.size();
  std::cout << "Species (" << nspecies << "): ";
  for (auto const& s : species) std::cout << s << " ";
  std::cout << std::endl;

  auto conc = torch::ones({1, nspecies}, torch::device(device).dtype(dtype)) *
              1e18;
  auto temp = 300.0 * torch::ones({1}, torch::device(device).dtype(dtype));
  auto pres = 1.0e5 * torch::ones({1}, torch::device(device).dtype(dtype));

  int nwave = 10;
  auto wave = torch::linspace(100, 300, nwave, torch::device(device).dtype(dtype));
  auto aflux = torch::ones({nwave}, torch::device(device).dtype(dtype)) * 1e14;

  std::map<std::string, torch::Tensor> extra;
  extra["wavelength"] = wave;
  extra["actinic_flux"] = aflux;

  auto [rate, rc_ddC, rc_ddT] = kinet->forward(temp, pres, conc, extra);

  std::cout << "Rate shape: " << rate.sizes() << std::endl;
  std::cout << "Rate: " << rate << std::endl;

  auto du = rate.matmul(kinet->stoich.t());
  std::cout << "du: " << du << std::endl;

  int nrxn = op->reactions().size();
  int n_reversible = 0;
  for (auto const& r : op->reactions()) {
    if (r.reversible()) n_reversible++;
  }
  EXPECT_EQ(rate.size(-1), (int64_t)(nrxn + n_reversible));
  EXPECT_TRUE(rate.isfinite().any().item<bool>());
}

int main(int argc, char** argv) {
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
